panynj.go

```diff
@@ -0,0 +1,203 @@
+	return &PaNyNjClient{httpClient: httoClient, clock: clock}
+}
+
+func (client *PaNyNjClient) GetTrainsAtStation(_ context.Context, station sourceapi.Station) ([]Train, error) {
+	realtimeApiContent, err := client.getContent()
+	if err != nil {
+		return nil, err
+	}
+	response := panynj.RidePathResponse{}
+	err = json.Unmarshal(realtimeApiContent, &response)
+	if err != nil {
+		return nil, err
+	}
+	var trains []Train
+	for _, result := range response.GetResults() {
+		consideredStation := client.convertStationAsStringToStation(result.GetConsideredStation())
+		if consideredStation != station {
+			continue
+		}
+		for _, destination := range result.GetDestinations() {
+			label := destination.GetLabel()
+			for _, message := range destination.GetMessages() {
+				upcomingTrain := sourceapi.GetUpcomingTrainsResponse_UpcomingTrain{
+					Route:            client.convertLineColorToRoute(message.GetLineColor()),
+					Direction:        client.convertDirectionAsStringToDirection(label),
+					ProjectedArrival: client.convertApiSecondsToArrivalAsStringToTimestamp(message.GetSecondsToArrival()),
+					LastUpdated:      client.convertApiLastUpdatedTimeStringToTimestamp(message.GetLastUpdated()),
+				}
+				trains = append(trains, &upcomingTrain)
+			}
+		}
+	}
+	return trains, nil
+}
+
+func (client *PaNyNjClient) GetStationToStopId(_ context.Context) (map[sourceapi.Station]string, error) {
+	return sourceStationToGtfsStopId, nil
+}
+
+func (client *PaNyNjClient) GetRouteToRouteId(_ context.Context) (map[sourceapi.Route]string, error) {
+	return sourceRouteToGtfsRouteId, nil
+}
+
+func (client *PaNyNjClient) convertDirectionAsStringToDirection(directionAsString string) sourceapi.Direction {
+	direction, ok := panynjLabelToDirection[strings.ToUpper(directionAsString)]
+	if !ok {
+		return sourceapi.Direction_DIRECTION_UNSPECIFIED
+	}
+	return direction
+}
+
+func (client *PaNyNjClient) convertStationAsStringToStation(stationAsString string) sourceapi.Station {
+	station, ok := panynjStationToSourceStation[stationAsString]
+	if !ok {
+		return sourceapi.Station_STATION_UNSPECIFIED
+	}
+	return station
+}
+
+func (client *PaNyNjClient) convertLineColorToRoute(lineColor string) sourceapi.Route {
+	route, ok := panynjLineColorToRoute[strings.ToUpper(lineColor)]
+	if !ok {
+		return sourceapi.Route_ROUTE_UNSPECIFIED
+	}
+	return route
+}
+
+func (client *PaNyNjClient) convertApiLastUpdatedTimeStringToTimestamp(timeString string) *timestamp.Timestamp {
+	const layout = "2006-01-02T15:04:05.000000-07:00"
+	timeObj, err := time.Parse(layout, timeString)
+	if err != nil {
+		return nil
+	}
+	value := timestamp.Timestamp{Seconds: timeObj.Unix()}
+	return &value
+}
+
+func (client *PaNyNjClient) convertApiSecondsToArrivalAsStringToTimestamp(secondsToArrivalAsString string) *timestamp.Timestamp {
+	secondsToArrival, err := strconv.ParseInt(secondsToArrivalAsString, 10, 64)
+	if err != nil {
+		return nil
+	}
+	nowSeconds := client.clock.Now().Unix()
```

Comment:
Would it make sense to calculate the arrival time as (secondsToArrival+lastUpdated)? This would potentially be more accurate and would also make the GTFS output not depend on the time the code runs.

Reply:
Good catch, I think this is much better for accuracy! I've updated the code to use `lastUpdated`.